core/types/subtyping.cc

```diff
@@ -1472,8 +1474,29 @@ bool Types::isSubTypeUnderConstraint(const GlobalState &gs, TypeConstraint &cons
     }
 
     if (auto *a2 = cast_type<AndType>(t2)) { // 2, 5
-        return Types::isSubTypeUnderConstraint(gs, constr, t1, a2->left, mode, errorDetailsCollector) &&
-               Types::isSubTypeUnderConstraint(gs, constr, t1, a2->right, mode, errorDetailsCollector);
+        auto subCollectorLeft = errorDetailsCollector.newCollector();
+        auto isSubTypeOfLeft = Types::isSubTypeUnderConstraint(gs, constr, t1, a2->left, mode, subCollectorLeft);
+        if (!isSubTypeOfLeft) {
+            if constexpr (shouldAddErrorDetails) {
+                auto message = ErrorColors::format("`{}` is not a subtype of `{}` (the left side of `{}`)", t1.show(gs),
+                                                   a2->left.show(gs), t2.show(gs));
+                subCollectorLeft.message = message;
+                errorDetailsCollector.addErrorDetails(move(subCollectorLeft));
+            }
+            return isSubTypeOfLeft;
+        }
+
+        auto subCollectorRight = errorDetailsCollector.newCollector();
+        auto isSubTypeOfRight = Types::isSubTypeUnderConstraint(gs, constr, t1, a2->right, mode, subCollectorRight);
+        if (!isSubTypeOfRight) {
+            if constexpr (shouldAddErrorDetails) {
+                auto message = ErrorColors::format("`{}` is not a subtype of `{}` (the right side of `{}`)",
+                                                   t1.show(gs), a2->right.show(gs), t2.show(gs));
+                subCollectorRight.message = message;
+                errorDetailsCollector.addErrorDetails(move(subCollectorRight));
+            }
+        }
+        return isSubTypeOfLeft && isSubTypeOfRight;
```

Comment:
Is this redundant? Can it be

```suggestion
        return isSubTypeOfRight;
```

?

Also, can we move the `if constexpr` for this half to enclose the `if (!isSubTypeOfRight)`, to make it easier on the optimizer?